### ISSUE TITLE
Replace use detection with a call graph

### ISSUE THREAD

**Issue by giacomocavalieri** [gleam-lang/gleam#2634]
I noticed that a private recursive function is not marked as unused.
For example this one raises no warnings:
```gleam
fn wobble() {
  wobble()
}
```

**Comment by lpil** [comment]
Yup! The current system isn't very good. I've been wanting to replace it with a proper call graph for ages but I've not got round to that yet.

I've not made an issue for it either, so let's use this one.

**Comment by giacomocavalieri** [comment]
Cool! Do you already have an idea on how to do that? I'm really curious to learn how that's going to be implemented 👀

**Comment by lpil** [comment]
Closes https://github.com/gleam-lang/gleam/issues/2579

**Comment by giacomocavalieri** [comment]
In this test the `unused` variable should be marked as unused!
```gleam
pub fn loop(cond: Bool, unused: Int) {
  case cond {
    True -> loop(False, unused)
    False -> 1
  }
}
```

**Comment by lpil** [comment]
I think it's fine not to detect that one. I tried with some other languages and they didn't.
